Add tests for DDP control flow models.

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#47206 Add tests for DDP control flow models.**

As discussed offline with @pritamdamania87, add testing to ensure per-iteration and rank-dependent control flow works as expected in DDP with `find_unused_parameters=True`.

Differential Revision: [D24659901](https://our.internmc.facebook.com/intern/diff/D24659901/)

rohan-varma committed:
Add tests for DDP control flow models.

As discussed offline with @pritamdamania87, add testing to ensure per-iteration and rank-dependent control flow works as expected in DDP with `find_unused_parameters=True`.

Differential Revision: [D24659901](https://our.internmc.facebook.com/intern/diff/D24659901/)

[ghstack-poisoned]

pritamdamania87 commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -3992,3 +3992,93 @@ def forward(_self, input, expected_type):  # noqa
+                    self.lin2 = nn.Linear(10, 10, bias=False)
+                    self.counter = 0
+
+                def forward(self, x):
```
Ideally we should test input dependent control flows (conditionals on x).

rohan-varma replied:
Updated to test based on `x`, basically we do a `torch.equal` using x and an expected tensor. Please let me know if there is a better way to test that.

pritamdamania87 commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -3992,3 +3992,93 @@ def forward(_self, input, expected_type):  # noqa
+                    self.counter = 0
+                    self.rank = rank
+
+                def forward(self, x):
```
same here, control flow should be on `x`

pritamdamania87 commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -3992,3 +3992,93 @@ def forward(_self, input, expected_type):  # noqa
+
+            torch.cuda.set_device(self.rank)
+            model = torch.nn.parallel.DistributedDataParallel(
+                ToyModel(self.rank).cuda(self.rank), device_ids=[self.rank], find_unused_parameters=True
```
Can we also add a test that verifies we raise an appropriate error when `find_unused_parameters=False`?

rohan-varma replied:
Added this to both the tests.

rohan-varma committed:
Update on "Add tests for DDP control flow models."


As discussed offline with @pritamdamania87, add testing to ensure per-iteration and rank-dependent control flow works as expected in DDP with `find_unused_parameters=True`.

Differential Revision: [D24659901](https://our.internmc.facebook.com/intern/diff/D24659901/)

[ghstack-poisoned]

zou3519 commented:
The tests introduced by this PR are failing: https://app.circleci.com/pipelines/github/pytorch/pytorch/235338/workflows/76ab99d0-701b-447e-8a67-7b0cc26ba7b5/jobs/8713716 so I'm reverting this

rohan-varma commented:
It looks like multigpu tests don't run on commits, and I tested only with a 2 GPU machine, which is why this was missed. Will resubmit with a ci-all.
